posthog/user_permissions.py

```diff
@@ -0,0 +1,266 @@
+            self._insight_permissions[insight.pk] = UserInsightPermissions(self, insight)
+        return self._insight_permissions[insight.pk]
+
+    @cached_property
+    def team_ids_visible_for_user(self) -> List[int]:
+        candidate_teams = Team.objects.filter(organization_id__in=self.organizations.keys()).only(
+            "pk", "organization_id", "access_control"
+        )
+
+        return [team.pk for team in candidate_teams if self.team(team).effective_membership_level is not None]
+
+    # Cached properties/functions for efficient lookups in other classes
+
+    @cached_property
+    def current_organization(self) -> Optional[Organization]:
+        if self._current_team is None:
+            raise ValueError("Cannot call .current_organization without passing current team to UsePermissions")
+        return self.get_organization(self._current_team.organization_id)
+
+    def get_organization(self, organization_id: UUID) -> Optional[Organization]:
+        return self.organizations.get(organization_id)
+
+    @cached_property
+    def organizations(self) -> Dict[UUID, Organization]:
+        return {member.organization_id: member.organization for member in self.organization_memberships.values()}
+
+    @cached_property
+    def organization_memberships(self) -> Dict[UUID, OrganizationMembership]:
+        memberships = OrganizationMembership.objects.filter(user=self.user).select_related("organization")
+        return {membership.organization_id: membership for membership in memberships}
+
+    @cached_property
+    def explicit_team_memberships(self) -> Dict[UUID, Any]:
+        try:
+            from ee.models import ExplicitTeamMembership
+        except ImportError:
+            return {}
+
+        memberships = ExplicitTeamMembership.objects.filter(
+            parent_membership_id__in=[membership.pk for membership in self.organization_memberships.values()]
+        ).only("parent_membership_id", "level")
+        return {membership.parent_membership_id: membership.level for membership in memberships}
+
+    @cached_property
+    def dashboard_privileges(self) -> Dict[int, Dashboard.PrivilegeLevel]:
+        try:
+            from ee.models import DashboardPrivilege
+
+            rows = DashboardPrivilege.objects.filter(user=self.user).values_list("dashboard_id", "level")
+            return {dashboard_id: cast(Dashboard.PrivilegeLevel, level) for dashboard_id, level in rows}
+        except ImportError:
+            return {}
+
+    def set_preloaded_dashboard_tiles(self, tiles: List[DashboardTile]):
+        """
+        Allows for speeding up insight-related permissions code
+        """
+        self._tiles = tiles
+
+    @cached_property
+    def preloaded_insight_dashboards(self) -> Optional[List[Dashboard]]:
+        if self._tiles is None:
+            return None
+
+        dashboard_ids = set(tile.dashboard_id for tile in self._tiles)
+        return list(Dashboard.objects.filter(pk__in=dashboard_ids))
```

Comment:
Should we be excluding deleted dashboards and deleted tiles here?

I'm regretting the complexity that the soft deletion adds but....